examples/ex6p.cpp

```diff
@@ -280,32 +299,50 @@ int main(int argc, char *argv[])
          break;
       }
 
-      // 21. Update the finite element space (recalculate the number of DOFs,
+      // 22. Update the finite element space (recalculate the number of DOFs,
       //     etc.) and create a grid function update matrix. Apply the matrix
       //     to any GridFunctions over the space. In this case, the update
       //     matrix is an interpolation matrix so the updated GridFunction will
       //     still represent the same function as before refinement.
       fespace.Update();
       x.Update();
 
-      // 22. Load balance the mesh, and update the space and solution. Currently
+      // 23. Load balance the mesh, and update the space and solution. Currently
       //     available only for nonconforming meshes.
-      if (pmesh.Nonconforming())
+      if (pmesh->Nonconforming())
       {
-         pmesh.Rebalance();
+         pmesh->Rebalance();
 
          // Update the space and the GridFunction. This time the update matrix
          // redistributes the GridFunction among the processors.
          fespace.Update();
          x.Update();
       }
 
-      // 23. Inform also the bilinear and linear forms that the space has
+      // 24. Inform also the bilinear and linear forms that the space has
       //     changed.
       a.Update();
       b.Update();
+
+      // 25. Save the current state of the mesh every 5 iterations. The
+      //     computation can be restarted from this point. Note that unlike in
+      //     visualization, we need to use the 'ParPrint' method to save all
+      //     internal parallel data structures.
+      if ((it + 1) % 5 == 0)
```

Comment:
An alternative is to use the condition: `if (it > 0 && it % 5 == 0)`. That way we'll get a checkpoints after iterations 5, 10, 15, ... but not for the initial mesh, `it == 0`.

Reply:
Hm, I don't know, the first checkpoint will saved after 6 iterations in this way, I'd rather keep it as it is...

Reply:
I now realize that this checkpoint is at the wrong place, in my opinion. The right place is to save the mesh and solution after the solve on the mesh, not before. That is what I assumed was happening but I had not checked.

Reply:
Sorry for the late reply. The reasoning was to allow the example to continue from a checkpoint exactly as it would have if it had continued normally, that's why the checkpoint is saved at the end of the iteration and when restarted, the for loop will just restart with the same data. Maybe that's not how this is normally done, but after all it's really just a sketch. Another reason to have it at the end is stylistic -- we probably don't want to clutter the main part of the code, since checkpointing is secondary here. But if there is a consensus, it's not a problem to move the checkpoint after step 20 in the example.

Reply:
I think you both have valid points.  In a realistic simulation code we would want to save the mesh and solution immediately after the solve as @v-dobrev states.  If AMR is part of the solution process this would mean immediately refining the mesh after a restart.  This would complicate the code slightly but in a realistic simulation code that would be acceptable.

However, this example code is a special case as @jakubcerveny suggests.  The main purpose is to demonstrate AMR.  We could therefore argue that the refinement taking place in step 21 is actually the "solution" that we are really interested in so it makes more sense to save the refined mesh when storing the checkpoint.

Because I view this as a special case, I vote to keep the code as it is.

Reply:
Good points Mark. You convinced me -- I also vote to keep the checkpoint code as is.

Reply:
Just for the record: I'm not convinced that this is a good choice.

However, since you all agree, let's proceed without changing this.